jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/HttpClientStreamTest.java

```diff
@@ -1193,6 +1197,98 @@ public boolean handle(Request request, org.eclipse.jetty.server.Response respons
         assertTrue(clientLatch.await(timeoutInSeconds, TimeUnit.SECONDS));
     }
 
+    @ParameterizedTest
+    @MethodSource("transports")
+    public void testUploadWithRetainedData(Transport transport) throws Exception
+    {
+        List<Content.Chunk> chunks = new ArrayList<>();
+
+        start(transport, new Handler.Abstract()
+        {
+            @Override
+            public boolean handle(Request request, org.eclipse.jetty.server.Response response, Callback callback)
+            {
+                Runnable retainer = new Runnable()
+                {
+                    @Override
+                    public void run()
+                    {
+                        while (true)
+                        {
+                            Content.Chunk chunk = request.read();
+                            if (chunk == null)
+                            {
+                                request.demand(this);
+                                return;
+                            }
+
+                            if (chunk.hasRemaining())
+                            {
+                                ByteBuffer byteBuffer = chunk.getByteBuffer();
+                                if (chunk.canRetain())
```

Comment:
It's a test, surely we can retain?

Reply:
I don't know? Do all the transports provide retainable chunks?